docs/proposals/resource_management.md

```diff
@@ -0,0 +1,313 @@
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
+
+## Design
+
+High level requirements for the design are as follows:
+ - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
+ - Support multiple cgroup managers - systemd, cgroupfs, etc.
+
+How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.<br>
+We propose the following hierarchy: we have separate top level cgroups for each Qos class. Further each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class. All containers brought up by the pod are nested under the pod's cgroup.<br>
+
+Examples in the next section.
+
+#### Quality of Service classes
+
+A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and Best-Effort, in decreasing order of priority.<br>
+We use the following denotation:<br>
+`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ Besteffort Qos`
+
+
+#### Guaranteed Cgroup ####
+
+Guaranteed pods will be placed at the `root` cgroup by default. This can be made configurable in the future.
+
+/G/cpu.shares = summation of cpu requests of all guaranteed containers
+/G/PodUID/cpu.quota = cpu request of Pod
+/G/m.limit_in_bytes = summation of memory limits of all guaranteed containers
+/G/PodUID/m.limit = memory limit of Pod <br>
+
+Example:
+We have two pods Pod1 and Pod2 having Pod Spec given below
+
+```yaml
+kind: Pod
+metadata:
+    name: Pod1
+spec:
+    containers:
+        name: foo
+            resources:
+                limits:
+                    cpu: 10m
+                    memory: 1Gi
+        name: bar
+            resources:
+                limits:
+                    cpu: 100m
+                    memory: 2Gi
+```
+
+```yaml
+kind: Pod
+metadata:
+    name: Pod2
+spec:
+    containers:
+        name: foo
+            resources:
+                limits:
+                    cpu: 20m
+                    memory: 2Gii
+```
+
+Pod1 and Pod2 are both classified as Guaranteed and are nested under the `/` cgroup.<br>
+
+/G/cpu.shares = 130m <br>
+/G/Pod1/cpu.quota = 110m <br>
+/G/Pod2/cpu.quota = 20m <br>
+/G/m.limit_in_bytes = 5Gi <br>
+/G/Pod1/m.limit = 3Gi <br>
+/G/Pod2/m.limit = 2Gi
+
+#### Burstable Cgroup ####
+
+We have the following resource parameters for the `Bu` cgroup.
+
+'/Bu/cpu.shares = summation of all cpu requests <br>
+/Bu/PodUID/cpu.quota = Pod Cpu Limit <br>
+/Bu/m.limit_in_bytes = Allocatable - summation of memory requests/limits of guaranteed pods <br>
```

Comment:
Would you suggest having it? We felt that soft limits don't really offer much gains during memory pressure.